sympy/matrices/matrices.py

```diff
@@ -2478,10 +2478,14 @@ def do(x):
             if len(args) == 3 and isinstance(args[2], Callable):
                 op = args[2]
                 flat_list = []
-                for i in range(rows):
+                rows_range = cols_range = [cls._sympify(i) for i in range(max(rows, cols))]
+                if rows < cols:
+                    rows_range = rows_range[:rows]
+                elif cols < rows:
+                    cols_range = cols_range[:cols]
+                for i in rows_range:
                     flat_list.extend(
-                        [cls._sympify(op(cls._sympify(i), cls._sympify(j)))
-                         for j in range(cols)])
+                        [cls._sympify(op(i, j)) for j in cols_range])
```

Comment:
Do you think a warning not to calculate non-Sympy values could eliminate the need of this additional `_sympify`? Any operation done with the Ints will result in an Expr ... unless the user purposely does something like `int(i/(j + 1))` instead of `Integer(i/(j + 1))` in the lambda.

Reply:
I have experimented with 
1. Passing SymPy integer to the callback and not sympifying the collected results.
```
  _     ._   __/__   _ _  _  _ _/_   Recorded: 15:59:55  Samples:  816
 /_//_/// /_\ / //_// / //_'/ //     Duration: 1.209     CPU time: 1.188
/   _/                      v3.0.1
```

2. Passing python integer to the callback and sympifying the collected results
```
  _     ._   __/__   _ _  _  _ _/_   Recorded: 15:56:25  Samples:  931
 /_//_/// /_\ / //_// / //_'/ //     Duration: 1.456     CPU time: 1.453
/   _/                      v3.0.1
```

And both gave more faster matrix creation than
```

  _     ._   __/__   _ _  _  _ _/_   Recorded: 16:08:17  Samples:  1423
 /_//_/// /_\ / //_// / //_'/ //     Duration: 2.078     CPU time: 2.062
/   _/                      v3.0.1
``` 

I think that the arguments passed to, or passed by the callback would not necessarily have to be sympified, if the operation has no precision issues.
But before writing the `flat_list`, the input types have to be sympified can be a must.

And the first example hadn't been tested with the type checking, so it would possibly add more time for the type checking. And I'm afraid that threading through all the sympy types would not be much different than using `sympify`

So, personally, I think it would be better to make a new argument to specify whether the indices should be python integer or sympy integer.

Reply:
I have experimented with 
1. Passing SymPy integer to the callback and not sympifying the collected results.
```
  _     ._   __/__   _ _  _  _ _/_   Recorded: 15:59:55  Samples:  816
 /_//_/// /_\ / //_// / //_'/ //     Duration: 1.209     CPU time: 1.188
/   _/                      v3.0.1
```

2. Passing python integer to the callback and sympifying the collected results
```
  _     ._   __/__   _ _  _  _ _/_   Recorded: 15:56:25  Samples:  931
 /_//_/// /_\ / //_// / //_'/ //     Duration: 1.456     CPU time: 1.453
/   _/                      v3.0.1
```

And both gave more faster matrix creation than
```

  _     ._   __/__   _ _  _  _ _/_   Recorded: 16:08:17  Samples:  1423
 /_//_/// /_\ / //_// / //_'/ //     Duration: 2.078     CPU time: 2.062
/   _/                      v3.0.1
``` 

I think that the arguments passed to, or passed by the callback would not necessarily have to be sympified, if the operation has no precision issues.
But before writing the `flat_list`, the input types have to be sympified can be a must.

And the first example hadn't been tested with the type checking, so it would possibly add more time for the type checking. And I'm afraid that threading through all the sympy types would not be much different than using `sympify`

So, personally, I think it would be better to make a new argument to specify whether the indices should be python integer or sympy integer.